arbnode/batch_poster.go

```diff
@@ -208,6 +212,84 @@ func NewBatchPoster(l1Reader *headerreader.HeaderReader, inbox *InboxTracker, st
 	return b, nil
 }
 
+// checkRevert checks blocks with number in range [from, to] whether they
+// contain reverted batch_poster transaction.
+func (b *BatchPoster) checkReverts(ctx context.Context, from, to *big.Int) (bool, error) {
+	if from.Cmp(to) == 1 {
+		return false, fmt.Errorf("wrong range, from: %v is more to: %v", from.Int64(), to.Int64())
+	}
+	one := big.NewInt(1)
+	for number := new(big.Int).Set(from); number.Cmp(to) != 1; number.Add(number, one) {
+		block, err := b.l1Reader.Client().BlockByNumber(ctx, number)
+		if err != nil {
+			return false, fmt.Errorf("getting block: %v by number: %w", number.Int64(), err)
+		}
+		for idx, tx := range block.Transactions() {
+			from, err := b.l1Reader.Client().TransactionSender(ctx, tx, block.Hash(), uint(idx))
+			if err != nil {
+				return false, fmt.Errorf("getting sender of transaction tx: %v, %w", tx.Hash(), err)
+			}
+			if bytes.Equal(from.Bytes(), b.dataPoster.From().Bytes()) {
+				r, err := b.l1Reader.Client().TransactionReceipt(ctx, tx.Hash())
+				if err != nil {
+					return false, fmt.Errorf("getting a receipt for transaction: %v, %w", tx.Hash(), err)
+				}
+				if r.Status == types.ReceiptStatusFailed {
+					return true, nil
+				}
+			}
+		}
+	}
+	return false, nil
+}
+
+// pollForReverts runs a gouroutine that listens to l1 block headers, checks
+// if any transaction made by batch poster was reverted.
+func (b *BatchPoster) pollForReverts(ctx context.Context) {
+	headerCh, unsubscribe := b.l1Reader.Subscribe(false)
+	defer unsubscribe()
+	var (
+		last *big.Int // number of last seen block
+		one  = big.NewInt(1)
+	)
+	for {
+		// Poll until:
+		// - L1 headers reader channel is closed, or
+		// - polling is through context, or
+		// - we see a transaction in the block from dataposter that was reverted.
+		select {
+		case h, closed := <-headerCh:
+			if closed {
+				log.Info("L1 headers channel has been closed")
+				return
+			}
+			// If this is the first block header, set last seen as number-1.
+			if last == nil {
+				last = new(big.Int).Set(h.Number)
+				last.Sub(last, one)
+			}
+			if h.Number.Int64()-last.Int64() > 100 {
+				log.Warn("Large gap between past seen: %v and current block number: %v, skipping check for reverts", last.Int64(), h.Number.Int64())
```

Comment:
This log line should be formatted as `"field", value` instead of the printf format. Also FYI you don't need to turn the header numbers into int64s to print them out. Their default string formatting works fine.

Reply:
Done.